### ISSUE TITLE
SQL IN Query is not working as expected

### ISSUE THREAD

**Issue by mustafaakin** [partiql/partiql-lang-kotlin#524]
Hello, we have the following query with IN clause that is returning empty results:

```
SELECT * FROM users WHERE name IN (SELECT name FROM roles)
```

However, the same query works in Postgres.  Alternatively following works, but of course, it's not same query semantically:

```
SELECT * FROM users u WHERE EXISTS (SELECT r.name FROM roles r WHERE r.name = u.name)
```

What are we doing/assuming that is wrong?

Sample code to make trying easier:

```kotlin
  val ion = IonSystemBuilder.standard().build()
  val pipeline = CompilerPipeline.standard(ion)

  val globalVariables = Bindings.buildLazyBindings<ExprValue> {
    addBinding("users") {
      val values = ion.loader.load(
        """
        {
        "name": "mustafa",
        "age": 30,
        }
        {
        "name": "ilteris",
        "age": 26,
        }
        {
        "name": "onur",
        "age": 24,
        }""".trimMargin()
      )
      pipeline.valueFactory.newFromIonValue(values)
    }
    addBinding("roles") {
      val values = ion.loader.load(
        """
        {
        "username": "mustafa",
        "admin": true,
        }
        {
        "name": "ilteris",
        "admin": false,
        }""".trimMargin()
      )
      pipeline.valueFactory.newFromIonValue(values)
    }
  }

  val query = "SELECT * FROM users WHERE name IN (SELECT name FROM roles)"
  val e = pipeline.compile(query)
  val session = EvaluationSession.build {
    globals(globalVariables)
  }

  val result = e.eval(session)
  for (exprValue in result) {
    println(exprValue)
  }
```

**Comment by lziq** [comment]
This query works `"SELECT * FROM users WHERE name IN (SELECT VALUE name FROM roles)"`. 

The sub-query you wrote `SELECT name FROM roles` outputs `<<{'name': 'mustafa'}, {'name': 'ilteris'}>>`, which is a bag of structs. The `name` in `users` is a string. That's why it returns you empty result.

**Comment by mustafaakin** [comment]
Yes, it works. Thanks a lot!. By the way, is this expected per SQL spec and what Postgres and other databases are doing a workaround?

**Comment by dlurton** [comment]
`SELECT VALUE` is a work-around, however the original query should work according to SQL-92.

**Comment by lziq** [comment]
I think it is probably not possible to make the original query work. 

As https://github.com/partiql/partiql-lang-kotlin/blob/main/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L703 suggests, for `x IN y`, the `IN` operator should check if `x` matches any element in `y`. 

If you want it work, then we have to modify the rule so it becomes: 
For `x IN y`, returns true if `x` matches any element in `y`, or if the element is a struct, returns true if `x` matches the struct's value. 

However, I think something like this might be too confusing. 

I think in other DB, the return value of `SELECT` can be considered as either table or value, which does not matter. However, it matters in PartiQL as we want to query nested data. So we have to use `SELECT` and `SELECT VALUE` to distinguish 2 cases.

**Comment by dlurton** [comment]
It is absolutely possible to make this query work if:

- For the right-hand side of `IN`, we coerce a bag of structs with a single field to a bag of values, or...
- Modify `IN` to work on a bag of structs with a single field.  

We do either today, which is why `SELECT VALUE` is required.

**Comment by lziq** [comment]
If SQL-92 compliance asks for it, we should make it work though.

**Comment by johnedquinn** [comment]
As an update, please see [my comment](https://github.com/partiql/partiql-lang-kotlin/pull/621#issuecomment-1147754213) about why I haven't merged the approved PR. Need to discuss with others on implications.
